src/preprocessing/passes/bool_to_bv.cpp

```diff
@@ -221,7 +215,157 @@ void BoolToBV::lowerNode(const TNode& n)
     default: break;
   }
 
+  // check if it's safe to lower or rebuild the node
+  // Note: might have to rebuild to keep changes to children, even if this node
+  // isn't being lowered
+
+  // it's safe to lower if all the children are bit-vectors
+  bool safe_to_lower =
+      (new_kind != k);  // don't need to lower at all if kind hasn't changed
+
+  // it's safe to rebuild if rebuilding doesn't change any of the types of the
+  // children
+  bool safe_to_rebuild = true;
+
+  for (const Node& nn : n)
+  {
+    if (safe_to_lower)
+    {
+      safe_to_lower = fromCache(nn).getType().isBitVector();
+    }
+    if (safe_to_rebuild)
+    {
+      safe_to_rebuild = (fromCache(nn).getType() == nn.getType());
+    }
+    // if it's already not safe to do either, stop checking
+    if (!safe_to_lower && !safe_to_rebuild)
```

Comment:
Good catch. I was surprised to discover that [GCC doesn't find this kind optimization on its own](https://godbolt.org/z/XUba0d).